File: android/Reguerta/app/src/main/java/com/reguerta/user/data/commitments/FirestoreSeasonalCommitmentRepository.kt
package com.reguerta.user.data.commitments

import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.Source
import com.reguerta.user.data.firestore.ReguertaFirestoreCollection
import com.reguerta.user.data.firestore.ReguertaFirestoreEnvironment
import com.reguerta.user.data.firestore.ReguertaFirestorePath
import com.reguerta.user.data.firestore.toRepositoryException
import com.reguerta.user.domain.RepositoryErrorKind
import com.reguerta.user.domain.RepositoryException
import com.reguerta.user.domain.access.Member
import com.reguerta.user.domain.commitments.SeasonalCommitment
import com.reguerta.user.domain.commitments.SeasonalCommitmentRepository
import com.reguerta.user.domain.commitments.loadActiveCommitmentsForMember
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext

private const val SeasonalCommitmentCanonicalUserField = "userId"
private val SeasonalCommitmentUserReadFields = listOf(
    SeasonalCommitmentCanonicalUserField,
    "memberId",
    "uid",
    "user",
    "member",
    "userRef",
    "memberRef",
    "userID",
    "memberID",
)
private val SeasonalCommitmentProductFields = listOf(
    "productId",
    "product",
    "productRef",
    "commonProductId",
    "itemId",
)
private val SeasonalCommitmentSeasonFields = listOf(
    "seasonKey",
    "season",
    "campaignKey",
    "commitmentSeason",
)
private val SeasonalCommitmentQtyFields = listOf(
    "fixedQty",
    "fixedQtyPerOfferedWeek",
    "fixedQtyPerWeek",
    "weeklyQty",
    "qty",
    "quantity",
)

class FirestoreSeasonalCommitmentRepository(
    private val firestore: FirebaseFirestore,
    private val environment: ReguertaFirestoreEnvironment? = null,
) : SeasonalCommitmentRepository {
    private val firestorePath = ReguertaFirestorePath(environment = environment)

    private val commitmentsCollectionPath: String
        get() = firestorePath.collectionPath(ReguertaFirestoreCollection.SEASONAL_COMMITMENTS)

    override suspend fun getActiveCommitmentsForUser(userId: String): List<SeasonalCommitment> = withContext(Dispatchers.IO) {
        val normalizedLookup = userId.trim().ifBlank { return@withContext emptyList() }
        try {
            getActiveCommitments(
                lookupKey = normalizedLookup,
                serverOnly = false,
            )
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "seasonalCommitments")
        }
    }

    internal suspend fun getActiveCommitmentsForMemberFromServer(
        member: Member,
    ): List<SeasonalCommitment> = withContext(Dispatchers.IO) {
        try {
            loadActiveCommitmentsForMember(member) { lookupKey ->
                getActiveCommitments(lookupKey = lookupKey, serverOnly = true)
            }
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "seasonalCommitments.server")
        }
    }

    private suspend fun getActiveCommitments(
        lookupKey: String,
        serverOnly: Boolean,
    ): List<SeasonalCommitment> {
        val normalizedLookup = lookupKey.trim().ifBlank { return emptyList() }
        val query = firestore.collection(commitmentsCollectionPath)
            .whereEqualTo(SeasonalCommitmentCanonicalUserField, normalizedLookup)
        val snapshot = if (serverOnly) {
            query.get(Source.SERVER).await()
        } else {
            query.get().await()
        }
        return snapshot.documents
            .map { document -> document.toSeasonalCommitment() }
            .filter { commitment -> commitment.userId.matchesLookupUserId(normalizedLookup) }
            .filter(SeasonalCommitment::active)
            .sortedWith(compareBy<SeasonalCommitment> { it.seasonKey }.thenBy { it.productId })
    }
}

private fun com.google.firebase.firestore.DocumentSnapshot.toSeasonalCommitment(): SeasonalCommitment {
    if (!exists()) invalidCommitmentDocument()
    return decodeSeasonalCommitmentDocument(id, data ?: invalidCommitmentDocument())
}

internal fun decodeSeasonalCommitmentDocument(
    documentId: String,
    data: Map<String, Any?>,
): SeasonalCommitment {
    val userId = data.firstNormalizedId(SeasonalCommitmentUserReadFields) ?: invalidCommitmentDocument()
    val productId = data.firstNormalizedId(SeasonalCommitmentProductFields) ?: invalidCommitmentDocument()
    val seasonKey = data.firstNormalizedId(SeasonalCommitmentSeasonFields) ?: invalidCommitmentDocument()
    val fixedQty = data.firstPositiveDouble(SeasonalCommitmentQtyFields) ?: invalidCommitmentDocument()
    return SeasonalCommitment(
        id = documentId,
        userId = userId,
        productId = productId,
        productNameHint = data.optionalNormalizedText("productName")
            ?: data.optionalNormalizedText("productDisplayName")
            ?: data.optionalNormalizedText("name"),
        seasonKey = seasonKey,
        fixedQtyPerOfferedWeek = fixedQty,
        active = data.optionalBoolean("active", default = true),
        createdAtMillis = data.optionalTimestampMillis("createdAt"),
        updatedAtMillis = data.optionalTimestampMillis("updatedAt"),
    )
}

private fun Map<String, Any?>.firstNormalizedId(fields: List<String>): String? =
    fields.firstNotNullOfOrNull { field ->
        val value = this[field] ?: return@firstNotNullOfOrNull null
        value.asNormalizedId() ?: invalidCommitmentDocument()
    }

private fun Map<String, Any?>.firstPositiveDouble(fields: List<String>): Double? =
    fields.firstNotNullOfOrNull { field ->
        val value = this[field] ?: return@firstNotNullOfOrNull null
        value.toPositiveDoubleOrNull() ?: invalidCommitmentDocument()
    }

private fun Any?.toPositiveDoubleOrNull(): Double? = when (this) {
    is Number -> this.toDouble()
    is String -> this.replace(",", ".").trim().toDoubleOrNull()
    else -> null
}?.takeIf { it > 0.0 }
    ?.takeIf(Double::isFinite)

private fun Any?.asNormalizedId(): String? = when (this) {
    is String -> this.trim().ifBlank { null }?.let(::normalizePathLikeIdentifier)
    is DocumentReference -> this.id.trim().ifBlank { null }
    is Map<*, *> -> (
        this["id"]
            ?: this["documentId"]
            ?: this["documentID"]
            ?: this["path"]
        )
        .asNormalizedId()
    else -> null
}

private fun Any?.asNormalizedText(): String? = when (this) {
    is String -> this.trim().ifBlank { null }
    is Map<*, *> -> (
        this["name"]
            ?: this["displayName"]
            ?: this["title"]
        )
        .asNormalizedText()
    else -> null
}

private fun Map<String, Any?>.optionalNormalizedText(field: String): String? {
    val value = this[field] ?: return null
    return value.asNormalizedText() ?: invalidCommitmentDocument()
}

private fun Map<String, Any?>.optionalBoolean(field: String, default: Boolean): Boolean {
    val value = this[field] ?: return default
    if (value !is Boolean) invalidCommitmentDocument()
    return value
}

private fun Map<String, Any?>.optionalTimestampMillis(field: String): Long {
    val value = this[field] ?: return 0L
    if (value !is com.google.firebase.Timestamp) invalidCommitmentDocument()
    return value.toDate().time
}

private fun invalidCommitmentDocument(): Nothing = throw RepositoryException(
    kind = RepositoryErrorKind.INVALID_DATA,
    resource = "seasonalCommitments.document",
)

private fun normalizePathLikeIdentifier(raw: String): String {
    if (!raw.contains("/")) return raw
    val trailing = raw.substringAfterLast('/').trim()
    return trailing.ifBlank { raw }
}

private fun String.matchesLookupUserId(lookup: String): Boolean {
    val normalizedCurrent = normalizePathLikeIdentifier(trim())
    val normalizedLookup = normalizePathLikeIdentifier(lookup.trim())
    return normalizedCurrent == normalizedLookup || normalizedCurrent.equals(normalizedLookup, ignoreCase = true)
}

File: ios/Reguerta/Reguerta/Data/Commitments/FirestoreSeasonalCommitmentRepository.swift
import FirebaseCore
import FirebaseFirestore
import Foundation

private let seasonalCommitmentCanonicalUserField = "userId"
private let seasonalCommitmentUserReadFields = [
    seasonalCommitmentCanonicalUserField,
    "memberId",
    "uid",
    "user",
    "member",
    "userRef",
    "memberRef",
    "userID",
    "memberID"
]
private let seasonalCommitmentProductFields = [
    "productId",
    "product",
    "productRef",
    "commonProductId",
    "itemId"
]
private let seasonalCommitmentSeasonFields = [
    "seasonKey",
    "season",
    "campaignKey",
    "commitmentSeason"
]
private let seasonalCommitmentQtyFields = [
    "fixedQty",
    "fixedQtyPerOfferedWeek",
    "fixedQtyPerWeek",
    "weeklyQty",
    "qty",
    "quantity"
]

actor FirestoreSeasonalCommitmentRepository: SeasonalCommitmentRepository {
    private let storedDB: Firestore

    init(firebaseAppName: String) {
        guard let app = FirebaseApp.app(name: firebaseAppName) else {
            preconditionFailure("Firebase app is required for seasonal commitments")
        }
        self.storedDB = Firestore.firestore(app: app)
    }

    func activeCommitments(userId: String, environment: SessionEnvironment) async throws -> [SeasonalCommitment] {
        return try await activeCommitments(userId: userId, source: .defaultSource, environment: environment)
    }

    func activeCommitmentsFromServer(
        userId: String,
        environment: SessionEnvironment
    ) async throws -> [SeasonalCommitment] {
        return try await activeCommitments(userId: userId, source: .server, environment: environment)
    }

    private func activeCommitments(
        userId: String,
        source: SeasonalCommitmentReadSource,
        environment: SessionEnvironment
    ) async throws -> [SeasonalCommitment] {
        let normalizedLookup = userId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalizedLookup.isEmpty else { return [] }

        do {
            return try await canonicalDocuments(
                lookupValue: normalizedLookup,
                source: source,
                environment: environment
            )
                .map { document in
                    try Self.commitment(documentID: document.documentID, data: document.data())
                }
                .filter { $0.userId.matchesLookupUserId(normalizedLookup) }
                .filter(\.active)
                .sorted(by: Self.sortCommitments)
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: "seasonalCommitments")
        }
    }

    private func canonicalDocuments(
        lookupValue: String,
        source: SeasonalCommitmentReadSource,
        environment: SessionEnvironment
    ) async throws -> [QueryDocumentSnapshot] {
        let commitmentsCollection = storedDB.reguertaCollection(.seasonalCommitments, environment: environment)
        let query = commitmentsCollection.whereField(
            seasonalCommitmentCanonicalUserField,
            isEqualTo: lookupValue
        )
        try Task.checkCancellation()
        let snapshot = switch source {
        case .defaultSource:
            try await query.getDocuments()
        case .server:
            try await query.getDocuments(source: .server)
        }
        return snapshot.documents
    }

    private static func sortCommitments(_ lhs: SeasonalCommitment, _ rhs: SeasonalCommitment) -> Bool {
        if lhs.seasonKey.localizedCaseInsensitiveCompare(rhs.seasonKey) != .orderedSame {
            return lhs.seasonKey.localizedCaseInsensitiveCompare(rhs.seasonKey) == .orderedAscending
        }
        return lhs.productId.localizedCaseInsensitiveCompare(rhs.productId) == .orderedAscending
    }

    static func commitment(documentID: String, data: [String: Any]) throws -> SeasonalCommitment {
        guard let userId = try firstNormalizedID(in: data, fields: seasonalCommitmentUserReadFields),
              let productId = try firstNormalizedID(in: data, fields: seasonalCommitmentProductFields),
              let seasonKey = try firstNormalizedID(in: data, fields: seasonalCommitmentSeasonFields),
              let fixedQtyPerOfferedWeek = try firstPositiveDouble(in: data, fields: seasonalCommitmentQtyFields) else {
            throw invalidDocumentError
        }
        return SeasonalCommitment(
            id: documentID,
            userId: userId,
            productId: productId,
            productNameHint: try optionalNormalizedText(data, field: "productName") ??
                optionalNormalizedText(data, field: "productDisplayName") ??
                optionalNormalizedText(data, field: "name"),
            seasonKey: seasonKey,
            fixedQtyPerOfferedWeek: fixedQtyPerOfferedWeek,
            active: try optionalBool(data, field: "active", default: true),
            createdAtMillis: try optionalTimestampMillis(data, field: "createdAt"),
            updatedAtMillis: try optionalTimestampMillis(data, field: "updatedAt")
        )
    }

    private static func firstNormalizedID(in data: [String: Any], fields: [String]) throws -> String? {
        for field in fields {
            guard let value = data[field] else { continue }
            if value is NSNull { continue }
            guard let identifier = normalizedID(value) else { throw invalidDocumentError }
            return identifier
        }
        return nil
    }

    private static func firstPositiveDouble(in data: [String: Any], fields: [String]) throws -> Double? {
        for field in fields {
            guard let value = data[field] else { continue }
            if value is NSNull { continue }
            guard let number = positiveDouble(value) else { throw invalidDocumentError }
            return number
        }
        return nil
    }

    private static func normalizedString(_ value: Any?) -> String? {
        guard let string = value as? String else { return nil }
        let trimmed = string.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    private static func normalizedText(_ value: Any?) -> String? {
        if let string = normalizedString(value) {
            return string
        }
        if let dictionary = value as? [String: Any] {
            return normalizedText(dictionary["name"]) ??
                normalizedText(dictionary["displayName"]) ??
                normalizedText(dictionary["title"])
        }
        return nil
    }

    private static func optionalNormalizedText(_ data: [String: Any], field: String) throws -> String? {
        guard let value = data[field] else { return nil }
        if value is NSNull { return nil }
        guard let text = normalizedText(value) else { throw invalidDocumentError }
        return text
    }

    private static func normalizedID(_ value: Any?) -> String? {
        if let string = normalizedString(value) {
            return normalizePathLikeIdentifier(string)
        }
        if let reference = value as? DocumentReference {
            return normalizedString(reference.documentID)
        }
        if let dictionary = value as? [String: Any] {
            return normalizedID(dictionary["id"]) ??
                normalizedID(dictionary["documentId"]) ??
                normalizedID(dictionary["documentID"]) ??
                normalizedID(dictionary["path"])
        }
        return nil
    }

    fileprivate static func normalizePathLikeIdentifier(_ value: String) -> String {
        guard value.contains("/") else { return value }
        let trailing = value.split(separator: "/").last.map(String.init)?
            .trimmingCharacters(in: .whitespacesAndNewlines)
        return (trailing?.isEmpty == false) ? trailing! : value
    }

    private static func positiveDouble(_ value: Any?) -> Double? {
        if let number = value as? NSNumber,
           CFGetTypeID(number) != CFBooleanGetTypeID() {
            let double = number.doubleValue
            return double.isFinite && double > 0 ? double : nil
        }
        if let string = value as? String {
            let normalized = string
                .replacingOccurrences(of: ",", with: ".")
                .trimmingCharacters(in: .whitespacesAndNewlines)
            if let double = Double(normalized), double.isFinite, double > 0 {
                return double
            }
        }
        return nil
    }

    private static func optionalBool(_ data: [String: Any], field: String, default defaultValue: Bool) throws -> Bool {
        guard let value = data[field] else { return defaultValue }
        if value is NSNull { return defaultValue }
        guard let number = value as? NSNumber,
              CFGetTypeID(number) == CFBooleanGetTypeID() else {
            throw invalidDocumentError
        }
        return number.boolValue
    }

    private static func optionalTimestampMillis(_ data: [String: Any], field: String) throws -> Int64 {
        guard let value = data[field] else { return 0 }
        if value is NSNull { return 0 }
        guard let timestamp = value as? Timestamp else { throw invalidDocumentError }
        return Int64(timestamp.dateValue().timeIntervalSince1970 * 1_000)
    }

    private static var invalidDocumentError: RepositoryError {
        .invalidData(resource: "seasonalCommitments.document")
    }
}

private nonisolated enum SeasonalCommitmentReadSource: Sendable {
    case defaultSource
    case server
}

private extension String {
    func matchesLookupUserId(_ lookup: String) -> Bool {
        let current = FirestoreSeasonalCommitmentRepository.normalizePathLikeIdentifier(
            trimmingCharacters(in: .whitespacesAndNewlines)
        )
        let target = FirestoreSeasonalCommitmentRepository.normalizePathLikeIdentifier(
            lookup.trimmingCharacters(in: .whitespacesAndNewlines)
        )
        return current == target || current.caseInsensitiveCompare(target) == .orderedSame
    }
}
